### PR TITLE
Fix  issue where FilterQueryOption.ApplyTo requires ODataQueryContext.RequestContainer

### PR DESCRIPTION
**Pull request by robward-ms**
### Issues

This pull request fixes issue #1194, #1219 

### Description

Fix issue where FilterQueryOption.ApplyTo requires ODataQueryContext.RequestContainer.

This change allows FilterQueryOption to execute without setting a
RequestContainer on ODataQueryContext, which is an internal property
set by ODataQueryOptions from the requests' container.

In this stand-alone scenario, a container cannot be injected so the uses
of the container now have a fallback option for creating objects when the
container does not exist.

However, one of those objects was an internal assembly resolver. On
.NET, the default assembly resolver can be used. However, .NET Core
does not implementation a default assembly resolver so a new one has been
added to the .NET Core version in this change. It is a simple static
list of assemblies that should be used to resolve types used in the
EDM models.

The addition of a default assembly resolver fixes a known
issue where a NotImplemented() exception was thrown when no part
manager was found to fulfill the request for assemblies. The
default assembly resolver will now provide those assemblies.

To simply injecting of custom assemblies for EDM purposes, the
output of the part manager and the default assembly resolver are
combined into a distinct list of assemblies, allow you to put EDM 
classes into a separate assembly that does not reference ASP.NET Core.
By default, the part manager will add assemblies which reference ASP.NET Core.

To add a custom assembly to the default assembly resolver, use:

`DefaultAssembliesResolver.Assemblies.Add(typeof(<type from assembly>).Assembly);`

### Checklist (Uncheck if it is not completed)

- [X] *Test cases added*
- [X] *Build and test with one-click build and test script passed*

### Additional work necessary

Need to add docs for the default assembly resolver.

### CODE REVIEW

**Comment by robward-ms** [comment]
Turns out this little gem is supported in .NET Core 2.x (but not 1.x):

AppDomain.CurrentDomain.GetAssemblies();

That would allow the Edm model builder to have a really great fallback to getting assemblies outside of the ASP.NET Core context like the ASP.NET Classic version of WebApi did. I'm going to take advantage of that in a revision to this change, eliminating the DefaultAssemblyResolver.

File: src/Microsoft.AspNetCore.OData/Adapters/WebApiAssembliesResolver.cs [hunk 1]
```diff
@@ -44,14 +45,25 @@ public IEnumerable<Assembly> Assemblies
...
             get
             {
+                new List<Assembly>();
                 if (this.innerManager != null)
                 {
                     IList<ApplicationPart> parts = this.innerManager.ApplicationParts;
-                    return parts.Where(p => p is AssemblyPart).Select(p => (p as AssemblyPart).Assembly).Distinct();
+                    IEnumerable<Assembly> assemblies = parts
+                        .Where(p => p is AssemblyPart)
+                        .Select(p => (p as AssemblyPart).Assembly);
```
Comment (robward-ms) [inline src/Microsoft.AspNetCore.OData/Adapters/WebApiAssembliesResolver.cs]: Lost .Distinct(). Add it back.

File: src/Microsoft.AspNetCore.OData/Adapters/WebApiAssembliesResolver.cs [hunk 2]
```diff
@@ -47,11 +48,22 @@ public IEnumerable<Assembly> Assemblies
                 if (this.innerManager != null)
                 {
                     IList<ApplicationPart> parts = this.innerManager.ApplicationParts;
-                    return parts.Where(p => p is AssemblyPart).Select(p => (p as AssemblyPart).Assembly).Distinct();
+                    IEnumerable<Assembly> assemblies = parts
+                        .Where(p => p is AssemblyPart)
+                        .Select(p => (p as AssemblyPart).Assembly)
+                        .Distinct();
```
Comment (AlanWong-MS) [inline src/Microsoft.AspNetCore.OData/Adapters/WebApiAssembliesResolver.cs]: nit: use `OfType<AssemblyPart>` to save the need of doing a `p is AssemblyPart` check?

i.e.
```
IEnumerable<Assembly> assemblies = parts
    .OfType<AssemblyPart>()
    .Select(p => p.Assembly)
    .Distinct();
```

File: src/Microsoft.AspNet.OData.Shared/ODataQueryContext.cs [hunk 3]
```diff
@@ -136,6 +140,10 @@ public DefaultQuerySettings DefaultQuerySettings
         /// <summary>
         /// Gets the request container.
         /// </summary>
+        /// <remarks>
+        /// The services container may not be present. See the ocnstructor in this fail for
```
Comment (AlanWong-MS) [inline src/Microsoft.AspNet.OData.Shared/ODataQueryContext.cs]: typo: ocnstructor -> constructor

Comment (robward-ms) [reply src/Microsoft.AspNet.OData.Shared/ODataQueryContext.cs]: also: fail -> file.

**Pull request merged**